### ISSUE TITLE
New Model Architectures - Implementation and Documentation Details

### ISSUE THREAD

**Issue by jdsgomes** [pytorch/vision#5319]
### 🚀 The feature

When adding a new model architecture there are some design/implementation details and documentation requirements that need to be taken into account. This issue intents to track such details in a dynamic manner, as it is possible to change over time.

### Model development and training steps

When developing a new model there are some details not to be missed:

- Implement a model factory function for each of the model variants

- in the module constructor, [pass layer constructor instead of instance](https://github.com/pytorch/vision/blob/47bd962069ba03f753e7ba711cb825317be0b00a/torchvision/models/efficientnet.py#L88) for configurable layers like norm, activation, and log the api usage with `_log_api_usage_once(self)`

- fuse layers together with existing common blocks if possible; For example consecutive conv, bn, activation layers could be replaced by [ConvNormActication](https://github.com/pytorch/vision/blob/47bd962069ba03f753e7ba711cb825317be0b00a/torchvision/ops/misc.py#L104)

- define `__all__` in the beginning of the model file to expose model factory functions; import model public APIs (e.g. factory methods) in `torchvision/models/__init__.py`

- create the model builder using the new API and add it to the prototype area. Here is an [example](https://github.com/pytorch/vision/pull/4784/files) on how to do this. The new API requires adding more information about the weights such as the preprocessing transforms necessary for using the model, meta-data about the model, etc

- Make sure you write tests for the model itself (see `_check_input_backprop`, `_model_params` and `_model_params` in `test/test_models.py`) and for any new operators/transforms or important functions that you introduce

Note that this list is not exhaustive and there are details here related to the code quality etc, but these are rules that apply in all PRs (see [Contributing to TorchVision](https://github.com/pytorch/vision/blob/main/CONTRIBUTING.md)).

Once the model is implemented, you need to train the model using the reference scripts. For example, in order to train a classification resnet18 model you would:

1. go to `references/classification`

2. run the train command (for example `torchrun --nproc_per_node=8 train.py --model resnet18`)

After training the model, select the best checkpoint and estimate its accuracy with a batch size of 1 on a single GPU. This helps us get better measurements about the accuracy of the models and avoid variants introduced due to batch padding (read [here](https://github.com/pytorch/vision/pull/4609/commits/5264b1a670107bcb4dc89e83a369f6fd97466ef8) for more details).

Finally, run the model test to generate expected model files for testing. Please include those generated files in the PR as well.:

`EXPECTTEST_ACCEPT=1 pytest test/test_models.py -k {model_name}`

### Documentation and Pytorch Hub

- `docs/source/models.rst`:

    - add the model to the corresponding section (classification/detection/video etc.)

    - describe how to construct the model variants (with and without pre-trained weights)

    - add model metrics and reference to the original paper

- `hubconf.py`:

    - import the model factory functions

    - submit a PR to [https://github.com/pytorch/hub](https://github.com/pytorch/hub) with a model page (or update an existing one)

- `README.md` under the reference script folder:

    - command(s) to train the model

**Comment by datumbox** [comment]
I've pinned the issue for now but we should consider making use of the Wiki pages, which are better suited for this kind of content. PyTorch core uses them extensively, so it might be worth aligning.

**Comment by NicolasHug** [comment]
Are we expecting these guidelines to change often?
Otherwise we might also consider just having this as a `.md` file in the repo?

**Comment by jdsgomes** [comment]
> 

That is a good point, and has been discussed previously with @datumbox and also during this [PR review](https://github.com/pytorch/vision/pull/5315#discussion_r795151631). In short I think is fair so say that there are no strong feelings either way, but there were two main arguments to keep it in a ticket for now. First we didn't want to make the contribution guidelines too long, second the content can change. So I would still favour to keep it here for a while and if it seems stable enough we can move it to a `.md` file

**Comment by datumbox** [comment]
Things will eventually become stable. I think the biggest changes on this documentation will come from the following:
- The Multi-weight support API rollout
- The Model documentation revamping
- Potentially by the model testing revamping work

Once things stabilise we can move to `md` if you prefer.

**Comment by yassineAlouini** [comment]
Thanks for this great guide @jdsgomes. There is a small typo here: [ConvNormActication](https://github.com/pytorch/vision/blob/47bd962069ba03f753e7ba711cb825317be0b00a/torchvision/ops/misc.py#L104)
(should be Activation)
